packages/runtime/src/components/page/PageHead.tsx

```diff
@@ -0,0 +1,294 @@
+    if (typeof element === 'string') return element
+
+    if (!VALID_TAG_REGEX.test(element.type)) return null
+
+    const key = element.key ? `${snippet.id}:${element.key}` : snippet.id
+
+    return createElement(element.type, { ...element.props, key, 'data-makeswift-snippet': 'true' })
+  })
+}
+
+type Props = {
+  document: MakeswiftPageDocument
+}
+
+export function PageHead({ document: page }: Props): JSX.Element {
+  const { headSnippets } = usePageSnippets({ page })
+
+  const isInBuilder = useIsInBuilder()
+
+  const site = useCachedSite(isInBuilder ? page.site.id : null)
+  const baseLocalizedPage = page.localizedPages.find(({ parentId }) => parentId == null)
+
+  const favicon = page.meta.favicon ?? defaultFavicon
+  const title = baseLocalizedPage?.meta.title ?? page.meta.title
+  const description = baseLocalizedPage?.meta.description ?? page.meta.description
+  const keywords = baseLocalizedPage?.meta.keywords ?? page.meta.keywords
+  const socialImage = baseLocalizedPage?.meta.socialImage ?? page.meta.socialImage
+  const canonicalUrl = baseLocalizedPage?.seo.canonicalUrl ?? page.seo.canonicalUrl
+  const isIndexingBlocked = baseLocalizedPage?.seo.isIndexingBlocked ?? page.seo.isIndexingBlocked
+
+  const fontFamilyParamValue = useMemo(() => {
+    if (site == null) {
+      return page.fonts
+        .map(({ family, variants }) => {
+          return `${family.replace(/ /g, '+')}:${variants.join()}`
+        })
+        .join('|')
+    }
+
+    return site.googleFonts.edges
+      .filter((edge): edge is NonNullable<typeof edge> => edge != null)
+      .map(({ activeVariants, node: { family, variants } }) => {
+        const activeVariantSpecifiers = variants
+          .filter(variant =>
+            activeVariants.some(activeVariant => activeVariant.specifier === variant.specifier),
+          )
+          .map(variant => variant.specifier)
+          .join()
+
+        return `${family.replace(/ /g, '+')}:${activeVariantSpecifiers}`
+      })
+      .join('|')
+  }, [site, page])
+
+  // When the page snippets change, we run the cleanup function for any snippets
+  // that are no longer present. This process only happens in the builder
+  const previousHeadSnippets = useRef<MakeswiftPageDocument['snippets'] | null>(null)
+  useEffect(() => {
+    const headSnippetsToCleanUp = (previousHeadSnippets.current ?? [])
+      .filter(previousSnippet => previousSnippet.cleanup != null)
+      .filter(previousSnippet => !headSnippets.some(snippet => previousSnippet.id === snippet.id))
+
+    headSnippetsToCleanUp.forEach(snippetToCleanUp => {
+      if (snippetToCleanUp.cleanup == null) return
+
+      try {
+        const cleanUp = new Function(snippetToCleanUp.cleanup)
+        cleanUp()
+      } catch (err) {
+        console.error('Error attempting to parse snippet cleanup', err)
+        // Ignore errors from user input.
+      }
+    })
+
+    previousHeadSnippets.current = headSnippets
+  }, [headSnippets])
```

Comment:
We previously talked about this, we moved this behavior outside of the hook since this hook is really only for retrieving head snippets and body snippets (we call this hook in two places). We decided that registering this effect in the hook would unnecessarily it where it doesn't apply. Instead, we add this hook directly to the page head, where it should be executed.

Reply:
Yeah, that's right... it's very confusing to see these things separated.